File: 56-merge-intervals/56-merge-intervals.java
class Solution {
    public int[][] merge(int[][] intervals) {
        Arrays.sort(intervals,(a,b) -> Integer.compare(a[0],b[0]));
        List<int[]> res = new ArrayList<>();
        
        int[] prevInterval = intervals[0];
        for (int i=1; i<intervals.length; i++){
            int prevStart = prevInterval[0];
            int prevEnd = prevInterval[1];
            
            int currStart = intervals[i][0];
            int currEnd = intervals[i][1];
            
            // If the curr interval and prev interval is not overlapping, we add the prev interval to result.
            if(prevEnd<currStart){
                res.add(prevInterval);
                prevInterval = intervals[i];
            }
            
            // If they are overlapping, we will merge them and change our prev interval to merged interval
            else{
                prevInterval[0] = Math.min(prevStart,currStart);
                prevInterval[1] = Math.max(prevEnd,currEnd);
            }
        }
        
        res.add(prevInterval);
        
        int[][] result = new int[res.size()][2];
        for(int i=0;i<res.size();i++){
            result[i] = res.get(i);
        }
        
        return result;
        
    }
}

File: 56-merge-intervals/56-merge-intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        # sort according to the start time
        intervals.sort(key = lambda x:x[0])
        n = len(intervals)
        if n<=1:
            return intervals
        res = []
        res.append(intervals[0])
        
        for i in range(1,n):
            prev_start, prev_end = res[-1][0], res[-1][1]
            curr_start, curr_end = intervals[i][0], intervals[i][1]
            if(prev_end<curr_start):
                res.append(intervals[i])
            else:
                new_interval = [prev_start,max(prev_end,curr_end)]
                res.pop()
                res.append(new_interval)
        return res
